src/Pirouette/Term/FromPlutusIR.hs

```diff
@@ -130,18 +128,68 @@ trBindings bs = do
 bindingCtxDeps :: forall tyname name fun loc
                 . (PIRConstraint tyname name fun)
                => [(Name, PIR.Term tyname name DefaultUni fun loc)]
-               -> TrM loc (DepsOf Name)
-bindingCtxDeps termBinds = get >>= go termBinds
+               -> TrM loc ()
+bindingCtxDeps termBinds = do
+  deps  <- get
+  deps' <- go 32 termBinds deps M.empty
+  put deps'
   where
-    go :: [(Name, PIR.Term tyname name DefaultUni fun loc)]
+    -- The go function below runs a fixpoint computation on a delta over the original dependencies;
+    -- We keep computing a new delta until we find nothing different to add; just to be sure we
+    -- wont loop here, we decrease a gas-counter on every recursive call.
+    go :: Int
+       -> [(Name, PIR.Term tyname name DefaultUni fun loc)]
+       -> DepsOf Name
        -> DepsOf Name
        -> TrM loc (DepsOf Name)
-    go xs prevDeps = do
+    go i xs origDeps delta = do
       vs <- asks termStack
-      let newDeps = M.fromList $ map (second $ termCtxDeps vs prevDeps) xs
-      if prevDeps == newDeps
-      then return newDeps
-      else go xs newDeps
+      let currDeps = M.unionWith S.union origDeps delta
+      let newDelta = M.fromList $ map (second $ termCtxDeps vs currDeps) xs
+      when (i <= 0) $
+        throwError DependencyComputationOverflow
+      if newDelta == delta
+      then return currDeps
+      else go (i-1) xs origDeps newDelta
```

Comment:
I think the main difference is that now we're comparing large maps for equality at each step versus comparing small changesets at each step. The gas counter is there as a sanity check; I agree its not required and I'm happy to drop it if you'd like

Reply:
I do not like gas when it can be avoided. If you have the feeling that this might run forever and we will have hard time to identify that it originates from this function, you can let it, otherwise, I would advocate for removing it.

Reply:
I'll remove it then; it makes sense that the new deltas tend to be the empty map